Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+    // reloadViewItems.
+    BOOL hasMalformedRowChange = NO;
+    NSMutableSet<NSString *> *updatedItemSet = [NSMutableSet new];
+    NSMutableSet<NSString *> *updatedNeighborItemSet = [NSMutableSet new];
+    for (YapDatabaseViewRowChange *rowChange in rowChanges) {
+        switch (rowChange.type) {
+            case YapDatabaseViewChangeUpdate: {
+                YapCollectionKey *collectionKey = rowChange.collectionKey;
+                if (collectionKey.key) {
+                    id<ConversationViewItem> _Nullable viewItem = self.viewItemCache[collectionKey.key];
+                    if (viewItem) {
+                        [self reloadInteractionForViewItem:viewItem];
+                        [updatedItemSet addObject:viewItem.itemId];
+                        if (rowChange.changes == YapDatabaseViewChangedDependency) {
+                            [updatedNeighborItemSet addObject:viewItem.itemId];
+                        }
+                    } else {
+                        hasMalformedRowChange = YES;
+                    }
+                } else if (rowChange.indexPath && rowChange.originalIndex < self.viewItems.count) {
+                    // Do nothing, this is a pseudo-update generated due to
+                    // setCellDrawingDependencyOffsets.
+                    OWSAssertDebug(rowChange.changes == YapDatabaseViewChangedDependency);
+                } else {
+                    hasMalformedRowChange = YES;
+                }
+                break;
+            }
+            case YapDatabaseViewChangeDelete: {
+                // Discard cached view items after deletes.
+                YapCollectionKey *collectionKey = rowChange.collectionKey;
+                if (collectionKey.key) {
+                    [self.viewItemCache removeObjectForKey:collectionKey.key];
+                } else {
+                    hasMalformedRowChange = YES;
+                }
+                break;
+            }
+            default:
+                break;
+        }
+        if (hasMalformedRowChange) {
+            break;
+        }
+    }
+
+    if (hasMalformedRowChange) {
+        // These errors seems to be very rare; they can only be reproduced
+        // using the more extreme actions in the debug UI.
+        OWSFailDebug(@"hasMalformedRowChange");
+        // resetMappings will call delegate.conversationViewModelDidUpdate.
+        [self resetMappings];
+        return;
+    }
+
+    NSUInteger oldViewItemCount = self.viewItems.count;
+    if (![self reloadViewItems]) {
+        // These errors are rare.
+        OWSFailDebug(@"could not reload view items; hard resetting message mappings.");
+        // resetMappings will call delegate.conversationViewModelDidUpdate.
+        [self resetMappings];
+        return;
+    }
+
+    OWSLogVerbose(@"self.viewItems.count: %zd -> %zd", oldViewItemCount, self.viewItems.count);
+
+    [self updateViewWitholdItemIdList:oldItemIdList
```

Comment:
Here's the bulk of the new logic, where we create a incremental/diff update for the CVC.

Reply:
Done.